src/semanticlayertools/clustering/reports.py

```diff
@@ -0,0 +1,288 @@
+
+    Generate reports to describe the content for all found clusters above a
+    minimal size by collecting metadata for all publications in each cluster,
+    finding the top 20 authors and affiliations of authors involved in the
+    cluster publications, and running basic NMF topic modelling with N=20 and
+    N=50 topics (english language models are used!).
+    For each cluster a report file is written to the output path.
+
+    Input CSV filename is used to create the output folder in output path. For
+    each cluster above the limit, a subfolder is created to contain all metadata
+    for the cluster. The metadata files are assumed to be in JSONL format and
+    contain the year in the filename.
+
+    :param infile: Path to input CSV file containing information on nodeid, clusterid, and year
+    :type infile: str
+    :param metadatapath: Path to JSONL (JSON line) formated metadata files.
+    :type metadatapath: str
+    :param outpath: Path to create output folder in, foldername reflects input filename
+    :type outpath: str
+
+    :param textcolumn: The dataframe column of metadata containing textutal for topic modelling (default=title)
+    :type textcolumn: str
+    :param numberProc: Number of CPU the routine will use (default = all!)
+    :type numberProc: int
+    :param minClusterSize: The minimal cluster size, above which clusters are considered (default=1000)
+    :type minClusterSize: int
+    :param timerange: Time range to evalute clusters for (usefull for limiting computation time, default = (1945, 2005))
+    :type timerange: tuple
+    """
+
+    def __init__(
+        self, infile: str, metadatapath: str, outpath: str,
+        textcolumn: str = 'title',
+        authorColumnName: str = 'author',
+        affiliationColumnName: str = 'aff',
+        publicationIDcolumn: str = 'nodeID',
+        numberProc: int = num_processes,
+        minClusterSize: int = 1000,
+        timerange: tuple = (1945, 2005)
+    ):
+        """Constructor method"""
+        self.numberProc = numberProc
+        self.minClusterSize = minClusterSize
+        self.metadatapath = metadatapath
+        self.textcolumn = textcolumn
+        self.authorColumnName = authorColumnName
+        self.affiliationColumnName = affiliationColumnName
+        self.publicationIDcolumn = publicationIDcolumn
+        clusterdf = pd.read_csv(infile)
+        basedata = clusterdf.groupby(['year', 'cluster']).size().to_frame('counts').reset_index()
+        self.largeClusterList = list(
+            basedata.groupby('cluster').sum().query(f'counts > {self.minClusterSize}').index
+        )
+        self.clusternodes = clusterdf.query(
+            'cluster in @self.largeClusterList'
+        )
+        outfolder = infile.split(os.path.sep)[-1][:-4]
+        self.timerange = timerange
+        self.outpath = os.path.join(outpath, outfolder)
+        if os.path.isdir(self.outpath):
+            raise OSError(f'Output folder {self.outpath} exists. Aborting.')
```

Comment:
I guess you'll see the `OSError` vs `ValueError` comment later, I don't remember which file it was. But it applies here, too - I think `ValueError` is more suitable here.